diff -ruN a/mint/falcon.h b/mint/falcon.h
--- a/mint/falcon.h
+++ b/mint/falcon.h
@@ -33,7 +33,7 @@
 		lea	sp@(12),sp"					\
 	: "=r"(retvalue)						\
 	: "g"(n), "r"(_a), "r"(_b), "r"(_c), "r"(_d), "r"(_e)		\
-	: "d0", "d1", "d2", "a0", "a1", "a2", "memory");		\
+	: "d1", "d2", "a0", "a1", "a2", "memory");		\
 	retvalue;							\
 })
 #endif
@@ -59,7 +59,7 @@
 		lea	sp@(18),sp"					\
 	: "=r"(retvalue)						\
 	: "g"(n), "r"(_a), "r"(_b), "r"(_c), "r"(_d)			\
-	: "d0", "d1", "d2", "a0", "a1", "a2", "memory");		\
+	: "d1", "d2", "a0", "a1", "a2", "memory");		\
 	retvalue;							\
 })
 #endif
@@ -89,7 +89,7 @@
 		lea	sp@(26),sp"					\
 	: "=r"(retvalue)						\
 	: "g"(n), "r"(_a), "r"(_b), "r"(_c), "r"(_d), "r"(_e), "r"(_f)	\
-	: "d0", "d1", "d2", "a0", "a1", "a2", "memory");		\
+	: "d1", "d2", "a0", "a1", "a2", "memory");		\
 	retvalue;							\
 })
 #endif
@@ -111,7 +111,7 @@
 		lea	sp@(10),sp"					\
 	: "=r"(retvalue)						\
 	: "g"(n), "r"(_a), "r"(_b)					\
-	: "d0", "d1", "d2", "a0", "a1", "a2", "memory");		\
+	: "d1", "d2", "a0", "a1", "a2", "memory");		\
 	retvalue;							\
 })
 #endif
@@ -135,7 +135,7 @@
 		lea	sp@(12),sp"					\
 	: "=r"(retvalue)						\
 	: "g"(n), "r"(_a), "r"(_b), "r"(_c)				\
-	: "d0", "d1", "d2", "a0", "a1", "a2", "memory");		\
+	: "d1", "d2", "a0", "a1", "a2", "memory");		\
 	retvalue;							\
 })
 #endif
diff -ruN a/mint/linea.h b/mint/linea.h
--- a/mint/linea.h
+++ b/mint/linea.h
@@ -770,7 +770,7 @@
 		.word   0xA000"						\
 	: "=g"(__xaline), "=g"(__xfonts), "=g"(__xfuncs)  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 	__aline = __xaline;						\
 	__fonts = __xfonts;						\
@@ -785,7 +785,7 @@
 		.word   0xA001" \
 	: 						  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 })
 
@@ -797,7 +797,7 @@
 		.word   0xA002"   \
 	: "=g"(retvalue)				  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 	(int)retvalue;							\
 })
@@ -810,7 +810,7 @@
 		.word   0xA003" \
 	: 						  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 })
 
@@ -822,7 +822,7 @@
 		.word   0xA004" \
 	: 						  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 })
 
@@ -834,7 +834,7 @@
 		.word   0xA005" \
 	: 						  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 })
 
@@ -846,7 +846,7 @@
 		.word   0xA006" \
 	: 						  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 })
 
@@ -861,7 +861,7 @@
 		movml	sp@+, d2/a2/a6" \
 	: 						  /* outputs */	\
 	: "r"(P)					  /* inputs  */	\
-	: "d0", "d1", "a0", "a1"	  	/* clobbered regs */	\
+	: "d1", "a0", "a1"	  	/* clobbered regs */	\
 	);								\
 })
 
@@ -873,7 +873,7 @@
 		.word   0xA008" \
 	: 						  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 })
 
@@ -885,7 +885,7 @@
 		.word   0xA009" \
 	: 						  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 })
 
@@ -897,7 +897,7 @@
 		.word   0xA00A" \
 	: 						  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 })
 
@@ -909,7 +909,7 @@
 		.word   0xA00B" \
 	: 						  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 })
 
@@ -924,7 +924,7 @@
 		movl	sp@+,a6"\
 	: 						  /* outputs */	\
 	: "g"(P)					  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"	   /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"	   /* clobbered regs */	\
 	);								\
 })
 
@@ -943,7 +943,7 @@
 		movl	sp@+,a6"\
 	: 						  /* outputs */	\
 	: "g"(x), "g"(y), "g"(sd), "g"(ss)		  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"	   /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"	   /* clobbered regs */	\
 	);								\
 })
 
@@ -955,7 +955,7 @@
 		.word   0xA00E" \
 	: 						  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 })
 
@@ -967,7 +967,7 @@
 		.word   0xA00F" \
 	: 						  /* outputs */	\
 	: 						  /* inputs  */	\
-	: "d0", "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
+	: "d1", "d2", "a0", "a1", "a2"       /* clobbered regs */	\
 	);								\
 })
 
diff -ruN a/mint/osbind.h b/mint/osbind.h
--- a/mint/osbind.h
+++ b/mint/osbind.h
@@ -108,6 +108,8 @@
 #ifndef _MINT_OSBIND_H
 #define _MINT_OSBIND_H	1
 
+#define OSBIND_CLOBBER_LIST	"d1", "d2", "a0", "a1", "a2", "memory"
+
 #ifndef _FEATURES_H
 # include <features.h>
 #endif
diff -ruN a/sysvars.h b/sysvars.h
--- a/sysvars.h
+++ b/sysvars.h
@@ -0,0 +1 @@
+#include <mint/sysvars.h>
