--- a/mint/falcon.h
+++ b/mint/falcon.h
@@ -242,6 +242,8 @@
 		(short)(rez),(short)(mode))
 #define VsetMode(mode)							\
 	(short)trap_14_ww((short)88,(short)(mode))
+#define Vsetscreen VsetScreen
+#define Vsetmode VsetMode
 #define VgetMonitor()							\
 	(short)trap_14_w((short)89)
 #define	VgetSize(mode)							\
--- a/mint/osbind.h
+++ b/mint/osbind.h
@@ -108,6 +108,8 @@
 #ifndef _MINT_OSBIND_H
 #define _MINT_OSBIND_H	1
 
+#define OSBIND_CLOBBER_LIST "d1", "d2", "a0", "a1", "a2", "memory"
+
 #ifndef _FEATURES_H
 # include <features.h>
 #endif
--- a/mint/sysvars.h
+++ b/mint/sysvars.h
@@ -80,7 +80,7 @@
 #define _vbclock	((volatile unsigned long *) 0x462L) /* vbi counter         */
 #define _frclock	((volatile unsigned long *) 0x466L) /* #vbi not vblsem'ed  */
 
-#define _hz_200		((volatile unsigned long *) 0x4baL)
+#define _hz_200		((unsigned long *) 0x4baL)
 
 #define conterm		((char *) 0x484L)
 #define savptr		((long *) 0x4A2L)
--- a/string.h
+++ b/string.h
@@ -70,7 +70,7 @@
 			   const char* __src, size_t __n));
 
 /* Compare S1 and S2.  */
-__EXTERN int strcmp __P ((const char* __s1, const char* __s2)) __THROW __attribute_pure__ __nonnull ((1, 2));
+__EXTERN int strcmp __P ((const char* __s1, const char* __s2));
 /* Compare N characters of S1 and S2.  */
 __EXTERN int strncmp __P ((const char* __s1, const char* __s2, size_t __n));
 
@@ -180,7 +180,7 @@
 
 
 /* Return the length of S.  */
-__EXTERN size_t strlen __P ((const char* __s)) __THROW __attribute_pure__ __nonnull ((1));
+__EXTERN size_t strlen __P ((const char* __s));
 
 #ifdef	__USE_GNU
 /* Find the length of STRING, but scan at most MAXLEN characters.
@@ -337,88 +337,6 @@
 __EXTERN size_t strlcat __P ((char *dst, const char *src, size_t siz));
 
 
-/*
- * inline versions of some functions.
- */
-#ifdef __OPTIMIZE__
-extern __inline __attribute__((__gnu_inline__)) size_t strlen(const char *scan)
-{
-	const char *start = scan;
-
-	__asm__ __volatile__(
-		"1:\n"
-		" tst.b (%0)+\n"
-		" bne.s 1b"
-	: "+a"(scan)
-	:
-	: "cc");
-	return scan - start - 1;
-}
-#endif
-
-
-#ifdef __OPTIMIZE__
-extern __inline __attribute__((__gnu_inline__)) int strcmp(const char *s1, const char *s2)
-{
-#ifdef __mcoldfire__
-	unsigned long cmp;
-	unsigned long tmp;
-
-	__asm__ __volatile__(
-	   "1:\n"
-		" mvz.b 	(%[s2])+,%[tmp]\n"
-		" mvz.b 	(%[s1])+,%[cmp]\n"
-		" beq.s 	1f\n"
-		" cmp.l 	%[tmp],%[cmp]\n"
-		" beq.s 	1b\n"
-	   "1:\n"
-		" sub.l 	%[tmp],%[cmp]\n"
-	: [cmp]"=d"(cmp), [tmp]"=d"(tmp), [s1]"+a"(s1), [s2]"+a"(s2)
-	:
-	: "cc");
-	return cmp;
-#else
-	unsigned long cmp;
-	unsigned long tmp;
-
-	__asm__ __volatile__(
-		" moveq #0,%[tmp]\n"
-		" moveq #0,%[cmp]\n"
-	   "1:\n"
-		" move.b 	(%[s2])+,%[tmp]\n"
-		" move.b	(%[s1])+,%[cmp]\n"
-		" beq.s 	1f\n"
-		" cmp.b 	%[tmp],%[cmp]\n"
-		" beq.s 	1b\n"
-	   "1:\n"
-		" sub.l 	%[tmp],%[cmp]\n"
-	: [cmp]"=d"(cmp), [tmp]"=d"(tmp), [s1]"+a"(s1), [s2]"+a"(s2)
-	:
-	: "cc");
-	return cmp;
-#endif
-}
-#endif
-
-
-extern __inline __attribute__((__gnu_inline__)) char *__inline_strcpy(char *dest, const char *src);
-extern __inline __attribute__((__gnu_inline__)) char *__inline_strcpy(char *dest, const char *src)
-{
-	char *dscan = dest;
-
-	__asm__ __volatile__(
-	   "1:\n"
-		" move.b	(%[src])+,(%[dest])+\n"
-		" bne.s 	1b\n"
-	: [src]"+a"(src), [dest]"+a"(dscan)
-	:
-	: "cc" AND_MEMORY);
-	return dest;
-}
-#ifdef __OPTIMIZE__
-#define strcpy(dest, src) (__builtin_constant_p(src) && strlen(src) < 2 ? __builtin_strcpy(dest, src) : __inline_strcpy(dest, src))
-#endif
-
 __END_DECLS
 
 #endif /* _STRING_H */
--- a/sys/types.h
+++ b/sys/types.h
@@ -188,8 +188,6 @@
 /* In BSD <sys/types.h> is expected to define BYTE_ORDER.  */
 # include <endian.h>
 
-/* It also defines `fd_set' and the FD_* macros for `select'.  */
-# include <sys/select.h>
 
 /* BSD defines these symbols, so we follow.  */
 # include <sys/sysmacros.h>
--- a/sysvars.h
+++ b/sysvars.h
@@ -0,0 +1 @@
+#include <mint/sysvars.h>
