diff -Naur pciutils-3.1.7-linux/lib/internal.h pciutils-3.1.7-mint/lib/internal.h
--- pciutils-3.1.7-linux/lib/internal.h	2010-09-09 23:56:08.000000000 +0200
+++ pciutils-3.1.7-mint/lib/internal.h	2010-09-11 13:32:08.000000000 +0200
@@ -68,4 +68,4 @@
 
 extern struct pci_methods pm_intel_conf1, pm_intel_conf2, pm_linux_proc,
 	pm_fbsd_device, pm_aix_device, pm_nbsd_libpci, pm_obsd_device,
-	pm_dump, pm_linux_sysfs;
+	pm_dump, pm_linux_sysfs, pm_mint_pcibios;
